libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -364,14 +364,23 @@ void CodeTransform::operator()(Identifier const& _identifier)
 	if (m_scope->lookup(_identifier.name, Scope::NonconstVisitor(
 		[=](Scope::Variable& _var)
 		{
-			// TODO: opportunity for optimization: Do not DUP if this is the last reference
-			// to the top most element of the stack
-			if (int heightDiff = variableHeightDiff(_var, _identifier.name, false))
-				m_assembly.appendInstruction(dev::eth::dupInstruction(heightDiff));
+			decreaseReference(_identifier.name, _var);
+			if (int const heightDiff = variableHeightDiff(_var, _identifier.name, false))
+			{
+				if (m_variablesScheduledForDeletion.count(&_var))
+				{
+					if (heightDiff > 1)
+						m_assembly.appendInstruction(dev::eth::swapInstruction(heightDiff));
+
+					deleteVariable(_var);
```

Comment:
I think in case heightDiff is 1, the stack slot should not be marked as available inside deleteVariable.

Reply:
True. Same even if `heightDiff > 1`, right?

But still, replacing the `deleteVariable` line with

```
m_context->variableStackHeights.erase(&_var);
m_context->variableReferences.erase(&_var);
m_variablesScheduledForDeletion.erase(&_var);
```

I still get the exact same 24 test failures.

Reply:
I'm just commenting what I think is wrong in the code, not how to fix the tests :)

Reply:
Actually, I think you can only do this optimization if height diff is zero.